import_matrix_table needs to check that keys are in the schema

Konrad hit:
```  File "<decorator-gen-816>", line 2, in import_matrix_table
  File "/tmp/933dc754-b6e8-446f-986d-f4900becb0bd/hail-devel-ef2f48e719c3.zip/hail/utils/java.py", line 208, in handle_py4j
hail.utils.java.FatalError: AssertionError: assertion failed: WrappedArray(): Struct{f0:String,f1:Int32}

Java stack trace:
java.lang.AssertionError: assertion failed: WrappedArray(): Struct{f0:String,f1:Int32}
    at scala.Predef$.assert(Predef.scala:170)
    at is.hail.expr.types.MatrixType.<init>(MatrixType.scala:45)
    at is.hail.expr.types.MatrixType$.fromParts(MatrixType.scala:23)
    at is.hail.io.LoadMatrix$.apply(LoadMatrix.scala:343)
    at is.hail.HailContext$$anonfun$importMatrices$1.apply(HailContext.scala:555)
    at is.hail.HailContext$$anonfun$importMatrices$1.apply(HailContext.scala:555)
    at is.hail.HailContext.forceBGZip(HailContext.scala:498)
    at is.hail.HailContext.importMatrices(HailContext.scala:554)
    at is.hail.HailContext.importMatrix(HailContext.scala:540)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
    at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
    at py4j.Gateway.invoke(Gateway.java:280)
    at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
    at py4j.commands.CallCommand.execute(CallCommand.java:79)
    at py4j.GatewayConnection.run(GatewayConnection.java:214)
    at java.lang.Thread.run(Thread.java:748)
```

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -985,13 +985,13 @@ def import_table(paths, key=[], min_partitions=None, impute=False, no_header=Fal
 
 @typecheck(paths=oneof(str, listof(str)),
            row_fields=dictof(str, Type),
-           key=oneof(str, listof(str)),
+           row_key=oneof(str, listof(str)),
            entry_type=enumeration(tint32, tint64, tfloat32, tfloat64, tstr),
            missing=str,
            min_partitions=nullable(int),
            no_header=bool,
            force_bgz=bool)
-def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing="NA", min_partitions=None,
+def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, missing="NA", min_partitions=None,
                         no_header=False, force_bgz=False):
     """
     Import tab-delimited file(s) as a :class:`.MatrixTable`.
@@ -1042,7 +1042,6 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
     Partition key:
         'Barcode': str
     ----------------------------------------
->>>>>>> address comments
 
     If the header information is missing for the row fields, like in the following:
 
@@ -1080,17 +1079,22 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
     Notes
     -----
 
-    The structure of the imported matrix table is:
+    The resulting matrix table has the following structure:
 
-        - **row fields** are named as specified in the column header. If they
+        - The **row fields** are named as specified in the column header. If they
         are missing from the header or ``no_header=True``, row field names are
-        set to the strings `f0`, `f1`, ... (0-indexed) in column order.
-
-        - **col_id** is the column key. It contains the IDs from the header of
-        the imported file as a :obj:`str`. If ``no_header=True``, column IDs are
-        set to integers 0, 1, ... (also 0-indexed) in column order.
-
-        - **x** is the entry field that contains the data from the imported
+        set to the strings `f0`, `f1`, ... (0-indexed) in column order. The types
+        of all row fields must be specified in the `row_fields` argument.
+        - The **row key** is taken from the `row_key` argument, and must be a
+        subset of row fields. If left empty, the row key will be a new row field
+        `row_idx` of type :obj:`int`, whose values 0, 1, ... are an index on the
+        original rows of the matrix.
+        - There is one column field, **col_id**, which is a key field of type
+        :obj:str or :obj:int. By default, its values are the strings given by
+        the corresponding column names in the header line. If ``no_header=True``,
+        column IDs are set to integers 0, 1, ... (also 0-indexed) in column
+        order.
+        - There is one entry field, **x**, that contains the data from the imported
         matrix.
 
 
@@ -1110,7 +1114,7 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
     row_fields: :obj:`dict` of :obj:`str` to :class:`.Type`
         Columns to take as row fields in the MatrixTable. They must be located
         before all entry columns.
-    key: :obj:`str` or :obj:`list` of :obj:`str`
+    row_key: :obj:`str` or :obj:`list` of :obj:`str`
         Key fields(s). If empty, creates an index `row_id` to use as key.
     entry_type: :class:`.Type`
         Type of entries in matrix table. Must be one of: :py:data:`.tint32`,

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -1019,6 +1019,30 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
     >>> matrix1 = hl.import_matrix_table('data/matrix1.tsv',
     ...                                  row_fields={'Barcode': hl.tstr, 'Tissue': hl.tstr, 'Days':hl.tfloat32},
     ...                                  key='Barcode')
+    >>> matrix1.describe()
+    ----------------------------------------
+    Global fields:
+        None
+    ----------------------------------------
+    Column fields:
+        'col_id': str
+    ----------------------------------------
+    Row fields:
+        'Barcode': str
+        'Tissue': str
+        'Days': float32
+    ----------------------------------------
+    Entry fields:
+        'x': int32
+    ----------------------------------------
+    Column key:
+        'col_id': str
+    Row key:
+        'Barcode': str
+    Partition key:
+        'Barcode': str
+    ----------------------------------------
+>>>>>>> address comments
 
     If the header information is missing for the row fields, like in the following:
 
@@ -1055,6 +1079,21 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
 
     Notes
     -----
+
+    The structure of the imported matrix table is:
+
+        - **row fields** are named as specified in the column header. If they
+        are missing from the header or ``no_header=True``, row field names are
+        set to the strings `f0`, `f1`, ... (0-indexed) in column order.
+
+        - **col_id** is the column key. It contains the IDs from the header of
+        the imported file as a :obj:`str`. If ``no_header=True``, column IDs are
+        set to integers 0, 1, ... (also 0-indexed) in column order.
+
+        - **x** is the entry field that contains the data from the imported
+        matrix.
+
+
     All columns to be imported as row fields must be at the start of the row.
 
     Unlike import_table, no type imputation is done so types must be specified
@@ -1063,14 +1102,6 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
     The header information for row fields is allowed to be missing, if the
     column IDs are present, but the header must then consist only of tab-delimited
     column IDs (no row field names).
-
-    If row field names are missing (or the no_header flag is ``True``), row fields
-    are imported in column-order as `f0`, `f1`, ... (0-indexed).
-
-    If the no_header flag is ``True``, column IDs will be imported as integers
-    `0`, `1`, ... (also 0-indexed).
-
-    Entries are imported into an entry-indexed field `x`.
 
     Parameters
     ----------

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -1145,12 +1145,12 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
         if v not in {tint32, tint64, tfloat32, tfloat64, tstr}:
             raise FatalError("""import_matrix_table expects field types to be one of: 
             'int32', 'int64', 'float32', 'float64', 'str': field {} had type '{}'""".format(repr(k), v))
-    key = wrap_to_list(key)
+    row_key = wrap_to_list(row_key)
     if entry_type not in {tint32, tint64, tfloat32, tfloat64, tstr}:
         raise FatalError("""import_matrix_table expects entry types to be one of: 
         'int32', 'int64', 'float32', 'float64', 'str': found '{}'""".format(entry_type))
 
-    jmt = Env.hc()._jhc.importMatrix(paths, jrow_fields, key, entry_type._jtype, missing, joption(min_partitions),
+    jmt = Env.hc()._jhc.importMatrix(paths, jrow_fields, row_key, entry_type._jtype, missing, joption(min_partitions),
                                      no_header, force_bgz)
     return MatrixTable(jmt)
 

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -1082,21 +1082,21 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
 
     The resulting matrix table has the following structure:
 
-        - The row fields are named as specified in the column header. If they
-        are missing from the header or ``no_header=True``, row field names are
-        set to the strings `f0`, `f1`, ... (0-indexed) in column order. The types
-        of all row fields must be specified in the `row_fields` argument.
-        - The row key is taken from the `row_key` argument, and must be a
-        subset of row fields. If left empty, the row key will be a new row field
-        `row_idx` of type :obj:`int`, whose values 0, 1, ... index the original
-        rows of the matrix.
-        - There is one column field, **col_id**, which is a key field of type
-        :obj:str or :obj:int. By default, its values are the strings given by
-        the corresponding column names in the header line. If ``no_header=True``,
-        column IDs are set to integers 0, 1, ... (also 0-indexed) in column
-        order.
-        - There is one entry field, **x**, that contains the data from the imported
-        matrix.
+        * The row fields are named as specified in the column header. If they
+          are missing from the header or ``no_header=True``, row field names are
+          set to the strings `f0`, `f1`, ... (0-indexed) in column order. The types
+          of all row fields must be specified in the `row_fields` argument.
+        * The row key is taken from the `row_key` argument, and must be a
+          subset of row fields. If left empty, the row key will be a new row field
+          `row_idx` of type :obj:`int`, whose values 0, 1, ... index the original
+          rows of the matrix.
+        * There is one column field, **col_id**, which is a key field of type
+          :obj:str or :obj:int. By default, its values are the strings given by
+          the corresponding column names in the header line. If ``no_header=True``,
+          column IDs are set to integers 0, 1, ... (also 0-indexed) in column
+          order.
+        * There is one entry field, **x**, that contains the data from the imported
+          matrix.
 
 
     All columns to be imported as row fields must be at the start of the row.

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -1043,7 +1043,8 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
         'Barcode': str
     ----------------------------------------
 
-    If the header information is missing for the row fields, like in the following:
+    In this example, the header information is missing for the row fields, but
+    the column IDs are still present:
 
     .. code-block:: text
 
@@ -1081,14 +1082,14 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
 
     The resulting matrix table has the following structure:
 
-        - The **row fields** are named as specified in the column header. If they
+        - The row fields are named as specified in the column header. If they
         are missing from the header or ``no_header=True``, row field names are
         set to the strings `f0`, `f1`, ... (0-indexed) in column order. The types
         of all row fields must be specified in the `row_fields` argument.
-        - The **row key** is taken from the `row_key` argument, and must be a
+        - The row key is taken from the `row_key` argument, and must be a
         subset of row fields. If left empty, the row key will be a new row field
-        `row_idx` of type :obj:`int`, whose values 0, 1, ... are an index on the
-        original rows of the matrix.
+        `row_idx` of type :obj:`int`, whose values 0, 1, ... index the original
+        rows of the matrix.
         - There is one column field, **col_id**, which is a key field of type
         :obj:str or :obj:int. By default, its values are the strings given by
         the corresponding column names in the header line. If ``no_header=True``,
@@ -1101,7 +1102,8 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
     All columns to be imported as row fields must be at the start of the row.
 
     Unlike import_table, no type imputation is done so types must be specified
-    for all columns that should be imported as row fields.
+    for all columns that should be imported as row fields. (The other columns are
+    imported as entries in the matrix.)
 
     The header information for row fields is allowed to be missing, if the
     column IDs are present, but the header must then consist only of tab-delimited

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -1018,7 +1018,7 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
 
     >>> matrix1 = hl.import_matrix_table('data/matrix1.tsv',
     ...                                  row_fields={'Barcode': hl.tstr, 'Tissue': hl.tstr, 'Days':hl.tfloat32},
-    ...                                  key='Barcode')
+    ...                                  row_key='Barcode')
     >>> matrix1.describe()
     ----------------------------------------
     Global fields:
@@ -1059,7 +1059,7 @@ def import_matrix_table(paths, row_fields={}, row_key=[], entry_type=tint32, mis
 
     >>> matrix2 = hl.import_matrix_table('data/matrix2.tsv',
     ...                                  row_fields={'f0': hl.tstr, 'f1': hl.tstr, 'f2':hl.tfloat32},
-    ...                                  key='f0')
+    ...                                  row_key='f0')
     >>> matrix2.rename({'f0': 'Barcode', 'f1': 'Tissue', 'f2': 'Days'})
 
     Sometimes, the header and row information is missing completely:

--- a/python/hail/methods/impex.py
+++ b/python/hail/methods/impex.py
@@ -1067,8 +1067,8 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
     If row field names are missing (or the no_header flag is ``True``), row fields
     are imported in column-order as `f0`, `f1`, ... (0-indexed).
 
-    If the no_header flag is ``True``, column IDs will be imported as `col0`,
-    `col1`, ... (also 0-indexed).
+    If the no_header flag is ``True``, column IDs will be imported as integers
+    `0`, `1`, ... (also 0-indexed).
 
     Entries are imported into an entry-indexed field `x`.
 
@@ -1091,8 +1091,7 @@ def import_matrix_table(paths, row_fields={}, key=[], entry_type=tint32, missing
         Minimum number of partitions.
     no_header: :obj:`bool`
         If ``True``, assume the file has no header and name the row fields `f0`,
-        `f1`, ... `fK` (0-indexed) and the column keys `col0`, `col1`, ...
-        `colN`.
+        `f1`, ... `fK` (0-indexed) and the column keys 0, 1, ... N.
     force_bgz : :obj:`bool`
         If ``True``, load **.gz** files as blocked gzip files, assuming
         that they were actually compressed using the BGZ codec.

--- a/python/hail/tests/test_methods.py
+++ b/python/hail/tests/test_methods.py
@@ -791,3 +791,16 @@ class Tests(unittest.TestCase):
         self.assertEqual(mt['Days']._indices, mt._row_indices)
         self.assertEqual(mt['col_id']._indices, mt._col_indices)
         self.assertEqual(mt['row_id']._indices, mt._row_indices)
+
+        mt.count()
+
+        row_fields = {'f0': hl.tstr, 'f1': hl.tstr, 'f2': hl.tfloat32}
+        hl.import_matrix_table(doctest_file('matrix2.tsv'),
+                               row_fields=row_fields, key=[]).count()
+        hl.import_matrix_table(doctest_file('matrix3.tsv'),
+                               row_fields=row_fields,
+                               no_header=True).count()
+        hl.import_matrix_table(doctest_file('matrix3.tsv'),
+                               row_fields=row_fields,
+                               no_header=True,
+                               key=[]).count()

--- a/python/hail/tests/test_methods.py
+++ b/python/hail/tests/test_methods.py
@@ -803,10 +803,10 @@ class Tests(unittest.TestCase):
         hl.import_matrix_table(doctest_file('matrix3.tsv'),
                                row_fields=row_fields,
                                no_header=True,
-                               key=[]).count()
+                               row_key=[]).count()
         self.assertRaises(hl.utils.FatalError,
                      hl.import_matrix_table,
                      doctest_file('matrix3.tsv'),
                      row_fields=row_fields,
                      no_header=True,
-                     key=['foo'])
+                     row_key=['foo'])

--- a/python/hail/tests/test_methods.py
+++ b/python/hail/tests/test_methods.py
@@ -804,3 +804,9 @@ class Tests(unittest.TestCase):
                                row_fields=row_fields,
                                no_header=True,
                                key=[]).count()
+        self.assertRaises(hl.utils.FatalError,
+                     hl.import_matrix_table,
+                     doctest_file('matrix3.tsv'),
+                     row_fields=row_fields,
+                     no_header=True,
+                     key=['foo'])

--- a/python/hail/tests/test_methods.py
+++ b/python/hail/tests/test_methods.py
@@ -796,7 +796,7 @@ class Tests(unittest.TestCase):
 
         row_fields = {'f0': hl.tstr, 'f1': hl.tstr, 'f2': hl.tfloat32}
         hl.import_matrix_table(doctest_file('matrix2.tsv'),
-                               row_fields=row_fields, key=[]).count()
+                               row_fields=row_fields, row_key=[]).count()
         hl.import_matrix_table(doctest_file('matrix3.tsv'),
                                row_fields=row_fields,
                                no_header=True).count()

--- a/src/main/scala/is/hail/io/LoadMatrix.scala
+++ b/src/main/scala/is/hail/io/LoadMatrix.scala
@@ -290,6 +290,9 @@ object LoadMatrix {
     val sc = hc.sc
     val hConf = hc.hadoopConf
 
+    if (files.isEmpty)
+      fatal("no files specified for import_matrix_table.")
+
     val (header1, nCols) = parseHeader(hConf, files.head, sep, nAnnotations, noHeader)
     val (rowFieldNames, colIDs) = splitHeader(header1, nAnnotations, nCols)
 

--- a/src/main/scala/is/hail/io/LoadMatrix.scala
+++ b/src/main/scala/is/hail/io/LoadMatrix.scala
@@ -215,13 +215,13 @@ object LoadMatrix {
     }
   }
 
-  def splitHeader(cols: Array[String], nRowFields: Int, nColIDs: Int): (Array[String], Array[String]) = {
+  def splitHeader(cols: Array[String], nRowFields: Int, nColIDs: Int): (Array[String], Array[_]) = {
     if (cols.length == nColIDs) {
       (Array.tabulate(nRowFields)(i => s"f$i"), cols)
     } else if (cols.length == nColIDs + nRowFields) {
       (cols.take(nRowFields), cols.drop(nRowFields))
     } else if (cols.isEmpty) {
-      (Array.tabulate(nRowFields)(i => s"f$i"), Array.tabulate(nColIDs)(i => s"col$i"))
+      (Array.tabulate(nRowFields)(i => s"f$i"), Array.range(0, nColIDs))
     } else
       fatal(
         s"""Expected file header to contain all $nColIDs column IDs and
@@ -297,7 +297,8 @@ object LoadMatrix {
 
     val header1Bc = sc.broadcast(header1)
 
-    LoadMatrix.warnDuplicates(colIDs)
+    if (!noHeader)
+      LoadMatrix.warnDuplicates(colIDs.asInstanceOf[Array[String]])
 
     val lines = sc.textFilesLines(files, nPartitions.getOrElse(sc.defaultMinPartitions))
 
@@ -340,9 +341,16 @@ object LoadMatrix {
         (Array("row_id"),TStruct("row_id" -> TInt64()) ++ rowFieldType)
       else (keyFields, rowFieldType)
 
+    if (!keyFields.forall(rowType.fieldNames.contains))
+      fatal(
+        s"""Some row keys not found in row schema:
+           |    '${ keyFields.filter(!rowType.fieldNames.contains(_)).mkString("'\n    '") }'
+         """.stripMargin
+      )
+
     val matrixType = MatrixType.fromParts(
       TStruct.empty(),
-      colType = TStruct("col_id" -> TString()),
+      colType = TStruct("col_id" -> (if (noHeader) TInt32() else TString())),
       colKey = Array("col_id"),
       rowType = rowType,
       rowKey = rowKey.toFastIndexedSeq,

--- a/src/test/scala/is/hail/io/ImportMatrixSuite.scala
+++ b/src/test/scala/is/hail/io/ImportMatrixSuite.scala
@@ -47,7 +47,8 @@ class ImportMatrixSuite extends SparkSuite {
       new java.util.HashMap[String, String](),
       new java.util.HashMap[String, String]())
 
-    renamed.copy2(colValues = Array.tabulate[Annotation](vsm.colValues.length){ i => Row("col" + i.toString) })
+    renamed.copy2(colType = TStruct("s" -> TInt32()),
+      colValues = Array.tabulate[Annotation](vsm.colValues.length){ i => Row(i) })
   }
 
   def renameColKeyField(vsm: MatrixTable): MatrixTable = {

